8204 : Keyboard traps in flamegraph view for search option

viragpurnam opened the pull request:
Focus Trapped Issue. Keyboard focus is trapped in three specific scenarios mentioned below:

1. Focus is trapped in JMC Agent Preset Manager

Select Window -> JMC Agent Preset Manager
Click on "Import Fils..."/ "Export File" button
Press Esc button when dialog is displayed
Press "Tab" button for multi times

2. Focus is trapped in Description of Template Manager

Select JVM Browser
Start Flight Recording
Template Manager
Click New
Click Edit
Focus to Description

3. Focus is trapped in search tool in Flame Graph  (Mostly on Windows)

Launch JDK Mission Control
Open any JFR file
Navigate to "Outline" tab
In Java Application, choose "Threads"
Move focus to search tool in Flame Graph

Solution:
1. PresetManagerPage.java : Setting the focus back to tableInspector.
2. TemplateEditPage.java : Added TraverseListener() for description field.
3. FlamegraphSwingView.java : Fix is not straight forward as SWT to Swing and back to SWT traversal is creating issue here. I have just implemented work around for this. I have added KeyListener() to all the swing components. Once it traverse through all the components, I am setting the Focus back to SWT.  With this there is no keyboard focus trap. 
 
These issues are kind of blocker with respect to accessibility. Could someone please review the changes.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JMC-8204](https://bugs.openjdk.org/browse/JMC-8204): Keyboard traps in flamegraph view for search option (**Bug** - P2)


### Reviewers
 * [Alex Macdonald](https://openjdk.org/census#aptmac) (@aptmac - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmc.git pull/554/head:pull/554` \
`$ git checkout pull/554`

Update a local copy of the PR: \
`$ git checkout pull/554` \
`$ git pull https://git.openjdk.org/jmc.git pull/554/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 554`

View PR using the GUI difftool: \
`$ git pr show -t 554`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmc/pull/554.diff">https://git.openjdk.org/jmc/pull/554.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmc/pull/554#issuecomment-1961134374)

bric3 reviewed (commented):
For the different cases reported:

1. I'm not sure what's the issue, in previous versions, the focus doesn't seem to be trapped. Pressing tab seem to cycle correctly on the buttons of the _JMC Agent Preset Manager_ dialog

2. Looks good

3. I can cycle out of the search text field, but I don't know on which SWT component the focus lands.

Note: I don't know if they are "quirks" on focus traversal depending on the OS, I'm using macOs.

aptmac commented:
I'll make a reminder to take a look at this on my Windows machine later tonight, but I did take a quick look at these changes in Linux..

> Focus is trapped in JMC Agent Preset Manager
> Select Window -> JMC Agent Preset Manager
> Click on "Import Fils..."/ "Export File" button
> Press Esc button when dialog is displayed
> Press "Tab" button for multi times

Exiting out of the Import Files dialog still doesn't return focus to the Preset Manager window.

> Focus is trapped in Description of Template Manager
> Select JVM Browser
> Start Flight Recording
> Template Manager
> Click New
> Click Edit
> Focus to Description

This one looks to be fixed.

> Focus is trapped in search tool in Flame Graph (Mostly on Windows)
> Launch JDK Mission Control
> Open any JFR file
> Navigate to "Outline" tab
> In Java Application, choose "Threads"
> Move focus to search tool in Flame Graph

Is the idea that you type something into the Flame Graph and then try to use Tab to cycle back through all the pages back to the Flame Graph page?

FWIW, on Linux regardless of which JFR Outline page I start on the Tab navigation seems to get stuck behind the Results tab. I can see there's a little white artifact that gets drawn so I think it's getting stuck here: 

![Screenshot from 2024-02-26 10-29-17](https://github.com/openjdk/jmc/assets/10425301/1223693b-f7d2-44b2-89d2-ab92350f26c4)

If I start at the Flame Graph page and repeatedly tab, the navigation goes:

1. Flame Graph search bar
2. ??? (Can't find the visual clue as to what it's active on, but it's not stuck)
3. Outline
4. Outline Minimize
5. Lock Page Structure
6. View Menu
7. Selected JFR Outline Page (i.e., Java Application/Threads/etc.)
8. Results tab
9. Results Minimize
10. **Stuck** behind Results tab

If I start for example on the Java Application Page instead of the Flame Graph, it still gets stuck in the same way as above but can jump straight to number 7.

viragpurnam commented:
> I'll make a reminder to take a look at this on my Windows machine later tonight, but I did take a quick look at these changes in Linux..
> 
> > Focus is trapped in JMC Agent Preset Manager
> > Select Window -> JMC Agent Preset Manager
> > Click on "Import Fils..."/ "Export File" button
> > Press Esc button when dialog is displayed
> > Press "Tab" button for multi times
> 
> Exiting out of the Import Files dialog still doesn't return focus to the Preset Manager window.
> 
> > Focus is trapped in Description of Template Manager
> > Select JVM Browser
> > Start Flight Recording
> > Template Manager
> > Click New
> > Click Edit
> > Focus to Description
> 
> This one looks to be fixed.
> 
> > Focus is trapped in search tool in Flame Graph (Mostly on Windows)
> > Launch JDK Mission Control
> > Open any JFR file
> > Navigate to "Outline" tab
> > In Java Application, choose "Threads"
> > Move focus to search tool in Flame Graph
> 
> Is the idea that you type something into the Flame Graph and then try to use Tab to cycle back through all the pages back to the Flame Graph page?
> 
> FWIW, on Linux regardless of which JFR Outline page I start on the Tab navigation seems to get stuck behind the Results tab. I can see there's a little white artifact that gets drawn so I think it's getting stuck here:
> 
> ![Screenshot from 2024-02-26 10-29-17](https://private-user-images.githubusercontent.com/10425301/307850892-1223693b-f7d2-44b2-89d2-ab92350f26c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5NjI0MzMsIm5iZiI6MTcwODk2MjEzMywicGF0aCI6Ii8xMDQyNTMwMS8zMDc4NTA4OTItMTIyMzY5M2ItZjdkMi00NGIyLTg5ZDItYWI5MjM1MGYyNmM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDE1NDIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhODAzOWZlOGMzNTJkNzYxOWVhZmIxODZlYjFjOTM1MTE4NGEzNjczMTA2NDIxZTBmODVmNTQ0NjBiYjMwNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DGW5043WIomt9cjIo0eSBTFhPt2fiJOXr_Jb5a8_qQ0)
> 
> If I start at the Flame Graph page and repeatedly tab, the navigation goes:
> 
> 1. Flame Graph search bar
> 2. ??? (Can't find the visual clue as to what it's active on, but it's not stuck)
> 3. Outline
> 4. Outline Minimize
> 5. Lock Page Structure
> 6. View Menu
> 7. Selected JFR Outline Page (i.e., Java Application/Threads/etc.)
> 8. Results tab
> 9. Results Minimize
> 10. **Stuck** behind Results tab
> 
> If I start for example on the Java Application Page instead of the Flame Graph, it still gets stuck in the same way as above but can jump straight to number 7.

Thanks @aptmac for the review and comments. It will be great if you could please review these changes on Windows as well.

aptmac commented:
> Thanks @aptmac for the review and comments. It will be great if you could please review these changes on Windows as well.

Gave it a try on Windows and it looks good on that front. I haven't dug through the code yet but do agree with the points that Brice mentioned.

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -557,8 +655,45 @@ private JComponent createSearchControl() {
+					e.consume();
+				}
+			}
+		});
```
Extracted a method and added comments.

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -557,8 +655,45 @@ private JComponent createSearchControl() {
+			@Override
+			public void keyPressed(KeyEvent e) {
+				if (e.getKeyCode() == KeyEvent.VK_TAB) {
+					if (e.getModifiersEx() > 0) {
```
Extracted a method and added comment.

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -499,14 +507,104 @@ public void createPartControl(Composite parent) {
+							Display.getDefault().syncExec(new Runnable() {
+								public void run() {
+									setTraverseAlready(!traverseAlready);
+									if (traverseAlready) {
```
Removed the setter method.

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -499,14 +507,104 @@ public void createPartControl(Composite parent) {
+			panel.setFocusable(true);
+			panel.setFocusTraversalKeysEnabled(true);
+
+			panel.addKeyListener(new KeyAdapter() {
```
Extracted a method and added comments.

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -499,14 +507,104 @@ public void createPartControl(Composite parent) {
+				flamegraphComponent.setFocusable(true);
+				flamegraphComponent.setFocusTraversalKeysEnabled(true);
+
+				flamegraphComponent.addKeyListener(new KeyAdapter() {
```
Added comments and extracted to a method.

viragpurnam commented:
Hi @aptmac, Thanks for your review.
Hi @bric3, I have addressed all the review points and committed the changes. Could you please review the updated code.

bric3 reviewed (commented):
LGTM albeit a few things to fix.

Not tested on either Linux or Windows.

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -509,102 +509,23 @@ public void createPartControl(Composite parent) {
-							});
+			// Adding focus traversal and key listener for main panel
+			setFocusTraversalProperties(panel);
+			addKeyListenerForBkwdFocustoSWT(panel);
```
Changed from addKeyListenerForBkwdFocustoSWT to addKeyListenerForBkwdFocus**T**oSWT

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -674,28 +635,76 @@ public void keyPressed(KeyEvent e) {
+			public void keyPressed(KeyEvent e) {
+				if (e.getKeyCode() == KeyEvent.VK_TAB) {
+					// Shift + TAB
+					if (e.getModifiersEx() > 0) {
```
@bric3, Thanks. I have changed as suggested by you.

viragpurnam commented:
Hi @bric3, I have done the required changes. Could you please re-review the changes. Thanks.

aptmac reviewed (commented):
Left a few minor things in the comments. Overall looks good.

I'm not sure if anyone else is running Linux, but my tab navigation is still stuck if the Outline tab is opened in JMC, regardless of the flameview being open or not. Would be nice to see if this is just me or not, otherwise this issue could be updated to be (mainly) Windows-focused. Scenario 2 looks to be fixed across all platforms though, which is nice.

bric3 commented on application/org.openjdk.jmc.console.agent/src/main/java/org/openjdk/jmc/console/agent/manager/wizards/PresetManagerPage.java:
```diff
@@ -206,13 +206,15 @@ protected void onImportFilesButtonSelected(IStructuredSelection selection) {
 				}
 
 				tableInspector.getViewer().refresh();
+				tableInspector.setFocus();
```
That was my original suggestion, but since there's `setFocus` (a single statement), but at the same time it's just a suggestion that does not bring as much value since the goal of `setFocus` is somewhat straightforward.

viragpurnam replied:
I have added setFocus to Import Button as well as Export Button. I have extracted that to a method now.

bric3 commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -557,12 +577,134 @@ private JComponent createSearchControl() {
+	private void setFocusBackToSWT() {
+		Display.getDefault().syncExec(new Runnable() {
+			public void run() {
+				IWorkbenchPage activePage = PlatformUI.getWorkbench().getWorkbenchWindows()[0].getActivePage();
```
Interesting, I think Eclipse workbench allows for multiple windows (or detached one), could that be an issue for `PlatformUI.getWorkbench().getWorkbenchWindows()[0]` ?

viragpurnam replied:
Idea is to get the active editor and we get the same with PlatformUI.getWorkbench().getActiveWorkbenchWindow().getActivePage() as well. I have done the required change.

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -557,12 +577,134 @@ private JComponent createSearchControl() {
+	/**
+	 * Adding key listener to transfer focus forward or backward based on 'TAB' or 'Shift + TAB'
+	 * 
+	 * @param comp
```
I have removed all @param

viragpurnam commented on application/org.openjdk.jmc.flightrecorder.flamegraph/src/main/java/org/openjdk/jmc/flightrecorder/flamegraph/views/FlamegraphSwingView.java:
```diff
@@ -689,7 +831,8 @@ private void saveFlamegraph() {
 			var fileName = fd.getFileName().toLowerCase();
-			// FIXME: FileDialog filterIndex returns -1 (https://bugs.eclipse.org/bugs/show_bug.cgi?id=546256)
+			// FIXME: FileDialog filterIndex returns -1
+			// (https://bugs.eclipse.org/bugs/show_bug.cgi?id=546256)
```
Currently I have only Mac and Windows. We can check on Linux later. (9.1.0)

viragpurnam commented:
/integrate
